CIP: reduce inflation (#249)
celestiaorg/CIPs: cips/cip-29.md

> | cip            | 29                                                                                                                                                          |

Reviewer: ```suggestion
| cip            | TBD                                                                                                                                                        |
```

Reviewer: nit: https://cips.celestia.org/cip-template.html

> Note: READ CIP-1 BEFORE USING THIS TEMPLATE! This is the suggested template for new CIPs. After you have filled in the requisite fields, please delete these comments. Note that an CIP number will be assigned by an editor. When opening a pull request to submit your CIP, please use an abbreviated title in the filename, cip-draft_title_abbrev.md. The title should be 44 characters or less. It should not repeat the CIP number in title, irrespective of the category.

| type           | Standards Track                                                                                                                                             |
| category       | Core                                                                                                                                                        |
| created        | 2025-02-04                                                                                                                                                  |

## **Abstract**

This CIP proposes to reduce both the Celestia inflation and disinflation by 33%. Celestia has a thriving ecosystem and already enjoys a significant share of the overall DA market. The current nominal inflation schedule also results in high dollar-value rewards for stakers. By "jumping" the inflation schedule forward, this proposal seeks to moderate total issuance, thereby reducing overall token dilution while still providing a yield that remains competitive to ensure the security of the network.

## **Motivation**

> Celestia launched in October of 2023, and since its launch there has been a high demand for posting blobs to the network, while other solutions have not been able to scale in a decentralized manner. However, TIA’s bonding has been continuously high (peaking at around 72% and currently at about 65%) meaning that stakers are overly incentivized and the system is overpaying for security.

Reviewer 2: > Blocking the CIP based on quantifying security means this CIP nor any other in the future will be passed because its subjective.

Sorry for the confusion. I'm not blocking the CIP on quantifying the goals. That's an optional suggestion. The only blocking comment was that the reference implementation looked broken. 

> What we do know now is that inflation on Celestia has been high comparative to other networks that have launched before and since Celestia's Launch.

That seems reasonable and accurate. This CIP could remove the line about "overpaying for security" and insert a line about how Celestia's inflation is high relative to others. It would help to include data on inflation for other networks.

> Celestia launched in October of 2023, and since its launch there has been a high demand for posting blobs to the network, while other solutions have not been able to scale in a decentralized manner. However, TIA’s bonding has been continuously high (peaking at around 72% and currently at about 65%) meaning that stakers are overly incentivized and the system is overpaying for security.

Reviewer 3: Why not inflation to reduce on 2%.
If Jerome Powel consider 2% acceptable, why not celestia to run based on this inflation level?

> Celestia launched in October of 2023, and since its launch there has been a high demand for posting blobs to the network, while other solutions have not been able to scale in a decentralized manner. However, TIA’s bonding has been continuously high (peaking at around 72% and currently at about 65%) meaning that stakers are overly incentivized and the system is overpaying for security.

Reviewer 4: Because Celestia is a roughly 1 year old proof of stake network. If you compare with past networks, ~5% is already quite low.

At a 33% reduction the inflation rate is 4.82% in the first year of the reduction, and the APR is roughly 7.39% if the bonding ratio stays the same. This is currently around 2.2 times the APR of Ethereum staking, but Celestia is a smaller and younger network so a higher security budget is justified.

## **Specification**

### **Overview**

1. Reduce Inflation: We reduce the current inflation by 33% in the next upgrade (v4)  
2. Reduce Disinflation Rate: We reduce the disinflation rate by 33%  
3. Reward Calculation:

   * The reward distribution mechanism remains unchanged—rewards are still allocated pro rata to stakers and validators.

   * The only modification is to the base inflation parameter and the disinflation parameter; both drop by 33% on the next upgrade leading to lower yet sustainable inflation immediately

Implementers MUST ensure:

* The chain upgrade process includes the new inflation parameters without disrupting block production.

> * The new schedule is included in the next voting proposal to reflect the updated inflation rates on-chain.

Reviewer 2: This probably meant to say

```suggestion
* The new schedule is included in the next major version release to reflect the updated inflation rates on-chain in the next app version (v4).
```

> * The new schedule is included in the next voting proposal to reflect the updated inflation rates on-chain.

Reviewer: lgtm, is this correct @tac0turtle ?

| Year    | Original Inflation | New Adjusted Inflation | Notes                                                                                                                        |
|:--------|:-------------------|:-----------------------|:-----------------------------------------------------------------------------------------------------------------------------|
| **0**   | 8.00               | 8.00                   | Genesis year, no change.                                                                                                     |
| **1**   | 7.20               | 7.20                   | First disinflation applied (10%).                                                                                            |
| **1.5** | 7.20               | 4.82                   | 33% drop in inflation applied, disinflation rate reduced to 6.7%. (This will be implemented in the next upgrade of Celestia) |

> | **2**   | 6.48               | 4.50                   | Regular annual disinflation applied (6.7%).                                                                                  |

Reviewer 2: I'm calculating different numbers for the adjusted inflation. Here are the existing constants:

```
initialInflation = .08
disinflation = .1
```

The "New Adjusted Inflation" column is decreasing both of those by 33% so

```
initialInflation = .0536
disinflation = .067
```

The formula to calculate inflation rate in a given year is: 

```
InitialInflationRate * ((1 - DisinflationRate) ^ YearsSinceGenesis)
```

let's skip year 0 and 1 because we can hard-code them to use the previous constants. For year 1.5

```
.0536 * (1 - .067) ^ 1 = 0.0500088
```

For year 2

```
.0536 * (1 - .067) ^ 2 = 0.04665821
```

which diverge from the values in the table.

Reviewer 4: I think it's the formula that might not be correct. I just double checked the table in a spreadsheet.

Reviewer 2: Yea my formula is incorrect. 

```
InitialInflationRate * ((1 - DisinflationRate) ^ YearsSinceGenesis)
```

 no longer works because this proposal modifies the inflation rate and disinflation rate for years > 1.5

Reviewer 2: Even if I use the reference implementation in this CIP, I get different numbers from the ones in the table. See https://github.com/celestiaorg/celestia-app/pull/4299

## **Backwards Compatibility**

No backward compatibility issues are anticipated. This parameter change follows the standard CIP process, and current node operators should not require any additional changes besides upgrading to the new binary which automatically implements the changes discussed in this CIP.

## **Test Cases**

1. Parameter Verification Test:

   * Ensure that when the updated inflation parameters are applied at Year 3, the on-chain inflation rate reflects the new values instead of the old schedule.

2. Reward Distribution Test:

   * Validate that block rewards are distributed correctly under the new inflation rates.

## **Reference Implementation**

Below is a pseudo-code snippet for updating the inflation parameters. Actual implementation may vary by client:

Proposed implementation:

```go
func (m Minter) CalculateInflationRate(ctx sdk.Context, genesis time.Time) sdk.Dec {
	years := yearsSinceGenesis(genesis, ctx.BlockTime())

	// Default inflation calculation
	inflationRate := InitialInflationRateAsDec().Mul(
		sdk.OneDec().Sub(DisinflationRateAsDec()).Power(uint64(years)),
	)

> 	// For AppVersion \> 3, adjust the inflation rate:
> 	if ctx.ConsensusParams().Version.AppVersion \> 3 {

Reviewer 2: [blocking] why are there escape characters in this code?

```suggestion
	// For AppVersion > 3, adjust the inflation rate:
	if ctx.ConsensusParams().Version.AppVersion > 3 {
```

		// First, reduce the current inflation rate by 33%

> 		inflationRate \= inflationRate.Mul(sdk.NewDecWithPrec(67, 2)) // 0.67 \= 67%
> 
> 		// Then, if we are in year two or later, apply a one-time disinflation of 6.7%
> 		if years \>= 2 {
> 			inflationRate \= inflationRate.Mul(sdk.OneDec().Sub(sdk.NewDecWithPrec(67, 3))) // 1 \- 0.067 \= 0.933
> 		}
> 	}

Reviewer 2: ```suggestion
		inflationRate = inflationRate.Mul(sdk.NewDecWithPrec(67, 2)) // 0.67 is 67%

		// Then, if we are in year two or later, apply a one-time disinflation of 6.7%
		if years >= 2 {
			inflationRate = inflationRate.Mul(sdk.OneDec().Sub(sdk.NewDecWithPrec(67, 3))) // 1 - 0.067 = 0.933
		}
	}
```

	// Ensure the inflation rate does not fall below the target inflation rate.
	if inflationRate.LT(TargetInflationRateAsDec()) {
		return TargetInflationRateAsDec()
	}

	return inflationRate
}
```

## **Alternative Approaches**

Alternatives that were considered.

### **Alternative 0: Do nothing**

> If we continue to do nothing, Celestia’s inflation schedule may not be competitive with other networks’ and TIA’s onchain usage may be limited.

Reviewer 2: This is phrased confusingly b/c the inflation schedule doesn't need to be competitive with other networks'. The yield earned from staking _may_ want to be competitive with other networks but if the inflation remains high then the yield earned from staking will be competitive. This proposal is advocating for reducing the inflation rate so that would reduce the staking yield which makes it less competitive with other networks'.

Can you rephrase if that's not what is meant here?

### **Alternative 1: Dynamic Inflation based on bonded ratio**

The Cosmos hub has a market dynamic inflation schedule/range where the inflation increases and decreases the further away current stake is from an ideal target amount of bonded stake.

Inflation \= CurrentInflation \+ (1 \- BondedRatio / GoalBonded) \* (InflationRateChange / BlocksPerYear)

Similarly, the Solana community is currently considering a similar approach but with vastly faster corrections than the Cosmos hub (see [SIMD-0228](https://github.com/solana-foundation/solana-improvement-documents/pull/228) for details).

> Generally, the dynamic approach is more complex and less predictable. Celestia's community has a strong commitment to base layer simplicity and minimalism. A fixed inflation schedule is simple and less complex, probably less controversial,  and the right answer to this point in time

Reviewer 2: [typo]

```suggestion
Generally, the dynamic approach is more complex and less predictable. Celestia's community has a strong commitment to base layer simplicity and minimalism. A fixed inflation schedule is simple and less complex, probably less controversial, and the right answer at this point in time
```

### **Alternative 2: Accelerate the Inflation faster (or slower)**

Why not apply a disinflation rate of 10% only, or why not even a disinflation rate of 50%? We believe that a 33% drop of inflation and disinflation (as you can see by the numbers) has a sufficiently significant effect on the inflation without deviating too much from the original schedule. For comparison, here are the numbers if we applied 50% instead or kept the original schedule:

![inflation change](./assets/cip-29/inflation_change.png)

> If we accelerate the schedule *slower*, the system continues overpaying for security and the change would not really move the needle as it would be too close to the current schedule. If we accelerate even faster, we risk hitting the target inflation of 1.5% too fast or generally risk lowering the inflation too much too early. Celestia mainnet is just about a year old and while there is already a thriving ecosystem, we want to err on the side of caution. About \~5% inflation seems ideal for a proof-of-stake system which is still early and whose tokenomics will play out in the future.

Reviewer 2: > About \~5% inflation seems ideal for a proof-of-stake system which is still early and whose tokenomics will play out in the future.

What is this based on?

* The chain upgrade process includes the new inflation parameters without disrupting block production.

* The new schedule is included in the next major version release to reflect the updated inflation rates on-chain in the next app version (v4).

## **Parameters**

Below is the illustrative table showing the original versus the accelerated schedule. The "New Adjusted Inflation" column is an example of how inflation might look if we reduce it by 33% starting in the next upgrade. Note that for simplicity, the table assumes the inflation drop is applied at year 1.5 but if this proposal gets accepted, it would be immediately applied with the next upgrade.

|    Year | Original Inflation | New Adjusted Inflation | Notes                                                                   |
|--------:|-------------------:|-----------------------:|:------------------------------------------------------------------------|
|   **0** |               8.00 |                   8.00 | Genesis year, no change.                                                |
|   **1** |               7.20 |                   7.20 |                                                                         |

> | **1.5** |               7.20 |                5.00088 | Reduce year 0 inflation by 33%, lowering the inflation rate to 5.00088% |

Reviewer 4: ```suggestion
| **1.5** |               7.20 |                5.00088 | Reduce year 0 inflation by 33% and additionally apply 6.7% disinfaltion, lowering the inflation rate to 5.00088% |
```

| **3**  | 5.83           | 16.66            | 11.66            | 8.97             |
| **4**  | 5.25           | 15.0             | 10.5             | 8.08             |
| **5**  | 4.72           | 13.49            | 9.44             | 7.26             |
| **6**  | 4.25           | 12.14            | 8.5              | 6.54             |
| **7**  | 3.83           | 10.94            | 7.66             | 5.89             |
| **8**  | 3.44           | 9.83             | 6.88             | 5.29             |
| **9**  | 3.1            | 8.86             | 6.2              | 4.77             |
| **10** | 2.79           | 7.97             | 5.58             | 4.29             |
| **11** | 2.51           | 7.17             | 5.02             | 3.86             |
| **12** | 2.26           | 6.46             | 4.52             | 3.48             |
| **13** | 2.03           | 5.8              | 4.06             | 3.12             |
| **14** | 1.83           | 5.23             | 3.66             | 2.82             |
| **15** | 1.65           | 4.71             | 3.3              | 2.54             |
| **16** | 1.5            | 4.29             | 3.0              | 2.31             |

> Updated schedule:

Reviewer 2: There are two updated schedules below. The second table has higher precision so let's drop the first one.

> | Year    | Inflation Rate | APR (35% Staked) | APR (50% Staked) | APR (65% Staked) |
> |:--------|:---------------|:-----------------|:-----------------|:-----------------|
> | **0**   | 8.0            | 22.86            | 16.0             | 12.31            |
> | **1**   | 7.2            | 20.57            | 14.4             | 11.08            |
> | **1.5** | 5.00           | 14.29            | 10.00            | 7.69             |
> | **2**   | 4.67           | 13.33            | 9.33             | 7.18             |
> | **3**   | 4.35           | 12.44            | 8.71             | 6.70             |
> | **4**   | 4.06           | 11.60            | 8.12             | 6.25             |
> | **5**   | 3.79           | 10.83            | 7.58             | 5.83             |
> | **6**   | 3.54           | 10.10            | 7.07             | 5.44             |
> | **7**   | 3.30           | 9.42             | 6.60             | 5.07             |
> | **8**   | 3.08           | 8.79             | 6.16             | 4.73             |
> | **9**   | 2.87           | 8.20             | 5.74             | 4.42             |
> | **10**  | 2.68           | 7.65             | 5.36             | 4.12             |
> | **11**  | 2.50           | 7.14             | 5.00             | 3.85             |
> | **12**  | 2.33           | 6.66             | 4.66             | 3.59             |
> | **13**  | 2.18           | 6.22             | 4.35             | 3.35             |
> | **14**  | 2.03           | 5.80             | 4.06             | 3.12             |
> | **15**  | 1.89           | 5.41             | 3.79             | 2.91             |
> | **16**  | 1.77           | 5.05             | 3.53             | 2.72             |

Reviewer 2: ```suggestion
```

	return disinflationRateCip29AsDec
}

// CalculateInflationRate returns the inflation rate for the current year depending on
// the current block height in context. The inflation rate is expected to
// decrease every year according to the schedule specified in the README.
func (m Minter) CalculateInflationRate(ctx sdk.Context, genesis time.Time) sdk.Dec {
	if ctx.ConsensusParams().Version.AppVersion <= 3 {
		return calculateInflationRatePreCip29(ctx, genesis)
	} else {
		return calculateInflationRatePostCip29(ctx, genesis)
	}
}

func calculateInflationRatePreCip29(ctx sdk.Context, genesis time.Time) sdk.Dec {
	years := yearsSinceGenesis(genesis, ctx.BlockTime())
	inflationRate := InitialInflationRateAsDec().Mul(sdk.OneDec().Sub(DisinflationRateAsDec()).Power(uint64(years)))

	if inflationRate.LT(TargetInflationRateAsDec()) {
		return TargetInflationRateAsDec()
	}
	return inflationRate
}

func calculateInflationRatePostCip29(ctx sdk.Context, genesis time.Time) sdk.Dec {
	if ctx.ConsensusParams().Version.AppVersion <= 3 {
		panic("calculateInflationRatePostCip29 should not be called with AppVersion <= 3")
	}

	years := yearsSinceGenesis(genesis, ctx.BlockTime())

> 	initialInflationRate := InitialInflationRateCip29AsDec().Mul(sdk.OneDec().Sub(DisinflationRateCip29AsDec()).Power(uint64(years)))
> 
> 	// Ensure the inflation rate does not fall below the target inflation rate.
> 	if initialInflationRate.LT(TargetInflationRateAsDec()) {
> 		return TargetInflationRateAsDec()
> 	}
> 
> 	return initialInflationRate

Reviewer 2: [nit] rename initialInflationRate => inflationRate

```suggestion
	inflationRate := InitialInflationRateCip29AsDec().Mul(sdk.OneDec().Sub(DisinflationRateCip29AsDec()).Power(uint64(years)))
	if inflationRate.LT(TargetInflationRateAsDec()) {
		return TargetInflationRateAsDec()
	}
	return inflationRate
```

## **Parameters**

Below is the illustrative table showing the original versus the accelerated schedule. The "New Adjusted Inflation" column is an example of how inflation might look if we reduce it by 33% starting in the next upgrade. Note that for simplicity, the table assumes the inflation drop is applied at year 1.5 but if this proposal gets accepted, it would be immediately applied with the next upgrade.

|    Year | Original Inflation | New Adjusted Inflation | Notes                                                                   |
|--------:|-------------------:|-----------------------:|:------------------------------------------------------------------------|
|   **0** |               8.00 |                   8.00 | Genesis year, no change.                                                |
|   **1** |               7.20 |                   7.20 |                                                                         |

> | **1.5** |               7.20 |                5.00088 | Reduce year 0 inflation by 33%, lowering the inflation rate to 5.00088% |

Reviewer 4: Shouldn't this be "Reduce year *1* inflation by 33%" still?

Reviewer 2: The revised proposal is to preserve this formula at all times:

```
InitialInflationRate * ((1 - DisinflationRate) ^ YearsSinceGenesis)
```
for versions 1 through 3:
```
InitialInflationRate = 0.08
DisinflationRate = 0.1
```
for versions 4 and up:
```
InitialInflationRateCip29 = 0.0536
DisinflationRateCip29 = 0.067
```

so you're right on how to calculate the year 1.5 inflation. For years after that, you can use the new constants with the original formula.
